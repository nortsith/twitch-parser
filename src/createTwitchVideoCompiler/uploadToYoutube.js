// @flow

import fs from 'fs';
import { google } from 'googleapis';

async function removeEmptyParameters(params) {
  const parameters = params;

  Object.keys(parameters).forEach((value) => {
    if (!parameters[value] || parameters[value] === 'undefined') {
      delete parameters[value];
    }
  });

  return parameters;
}

async function createResource(props) {
  const resource = {};
  const normalizedProps = props;

  Object.keys(props).forEach((p) => {
    const value = props[p];
    if (p && p.substr(-2, 2) === '[]') {
      const adjustedName = p.replace('[]', '');
      if (value) {
        normalizedProps[adjustedName] = value.split(',');
      }
      delete normalizedProps[p];
    }
  });

  Object.keys(normalizedProps).forEach((p) => {
    if (Object.prototype.hasOwnProperty.call(normalizedProps, p) && normalizedProps[p]) {
      const propArray = p.split('.');
      let ref = resource;
      for (let pa = 0; pa < propArray.length; pa += 1) {
        const key = propArray[pa];
        if (pa === propArray.length - 1) {
          ref[key] = normalizedProps[p];
        } else {
          ref[key] = ref[key] || {};
          ref = ref[key];
        }
      }
    }
  });

  return resource;
}

export default async function uploadToYoutube(
  authorization: {},
  properties: {
    language: string,
    description: string,
    tags: string,
    title: string,
    privacy: string,
    videoPath: string,
  },
): Promise<void> {
  const requestData = {
    params: { part: 'snippet,status' },
    properties: {
      'snippet.categoryId': '20',
      'snippet.defaultLanguage': properties.language,
      'snippet.description': properties.description,
      'snippet.tags[]': properties.tags,
      'snippet.title': properties.title,
      'status.embeddable': '',
      'status.license': '',
      'status.privacyStatus': properties.privacy,
      'status.publicStatsViewable': '',
    },
    mediaFilename: properties.videoPath,
  };

  const service = google.youtube('v3');
  const parameters = await removeEmptyParameters(requestData.params);
  // $FlowFixMe
  parameters.auth = authorization;
  // $FlowFixMe
  parameters.media = { body: fs.createReadStream(requestData.mediaFilename) };
  // $FlowFixMe
  parameters.notifySubscribers = false;
  // $FlowFixMe
  parameters.resource = await createResource(requestData.properties);

  return new Promise((resolve, reject) => {
    service.videos.insert(parameters, (err, data) => {
      if (err) {
        console.log(`The API returned an error: ${err}`);
        reject();
      }
      if (data) {
        resolve();
      }
      process.exit();
    });

    console.log('Uploading...');
  });
}
