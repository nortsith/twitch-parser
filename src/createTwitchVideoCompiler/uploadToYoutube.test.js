// @flow

import { createParameters, type Properties } from './uploadToYoutube';

jest.mock('googleapis', () => ({
  google: {},
}));

jest.mock('fs', () => ({
  createReadStream: () => null,
}));

test('createParameters', () => {
  const authorization = null;
  const properties: Properties = {
    language: 'en',
    description: 'desc',
    tags: ['tagOne', 'tagTwo'],
    title: 'title',
    privacy: 'private',
    videoPath: 'test',
  };

  const parameters = createParameters(authorization, properties);

  expect(parameters).toMatchSnapshot();
});
