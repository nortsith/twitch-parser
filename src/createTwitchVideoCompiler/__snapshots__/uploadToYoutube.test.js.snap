// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createParameters 1`] = `
Object {
  "auth": null,
  "media": Object {
    "body": null,
  },
  "notifySubscribers": false,
  "part": "snippet,status",
  "resource": Object {
    "snippet": Object {
      "categoryId": "20",
      "defaultLanguage": "en",
      "description": "desc",
      "tags": Array [
        "tagOne",
        "tagTwo",
      ],
      "title": "title",
    },
    "status": Object {
      "privacyStatus": "private",
    },
  },
}
`;
